### PR TITLE
Allow passing inputs_embeds instead of input_ids

### LINKED ISSUE #727: 'NoneType' object has no attribute 'shape'
### System Info

peft==0.5.0.dev0   python=3.8

### Who can help?

@pacman100  
在peft_model.py的1094行的代码中
batch_size = input_ids.shape[0]

如果用户直接输入inputs_embeds，而不是 input_ids，就会报错。建议改成
 batch_size = input_ids.shape[0] if input_ids else inputs_embeds.shape[0]

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import torch
from peft import LoraConfig, TaskType
from peft import PromptTuningConfig, PrefixTuningConfig, AdaptionPromptConfig
from peft import IA3Config, IA3Model
from peft import get_peft_model
from transformers import BertModel,BertConfig
config = BertConfig()
bert = BertModel(config)

peft_config = PrefixTuningConfig(task_type=TaskType.SEQ_2_SEQ_LM, inference_mode=False, num_virtual_tokens=20)

model = get_peft_model(bert, peft_config)
model.print_trainable_parameters()

inputs = torch.randn([1,5,768])
output = model(inputs_embeds=inputs)
```

### Expected behavior

 Just like transformers, I recommend that either input_ids or inputs_embeds should work，as follow:

batch_size = input_ids.shape[0] if input_ids else inputs_embeds.shape[0]

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Resolves #727

## Description

Right now, there is an issue with a few `PeftModelForXxx` classes when users pass only `inputs_embeds` but not `input_ids`. First of all, the batch size was being derived on `input_ids`, now it is derived from `inputs_embeds` instead if `input_ids` is `None`. Furthermore, in `PeftModelForCausalLM`, the `forward` calls to the base model was not passing the `inputs_embeds` along, which resulted in errors down the line. These issues have been fixed now.

## Open issues

During testing, I ran into a some problems.

1. For decoder models, the test would fail for `GPTBigCodeForCausalLM`. I'm not sure why, but that model is already excluded in other tests, so I just excluded it for this test too.

2. For feature extraction tests, the test would fail for prefix tuning. Not sure exactly why only here. I thought I only needed pass the `inputs_embeds` to the `self.base_model call` for prefix tuning, but then I got the error below, which is why I just excluded those tests for the time being.

> TypeError: DebertaV2Model.forward() got an unexpected keyword argument
'past_key_values'

3. Similarly, I wanted to add the tests to encoder-decoder models, but there I got the error:

> ValueError: If no `decoder_input_ids` or `decoder_inputs_embeds` are
passed, `input_ids` cannot be `None`. Please pass either `input_ids` or `decoder_input_ids` or `decoder_inputs_embeds`.

   So now, the tests are not run for encoder-decoder.

It would be great if someone else with more expertise could ensure that those cases work correctly.

### CODE REVIEW

**Review by younesbelkada — APPROVED** [review]
Looking great thanks a lot! 
Regarding gptbigcode + prompt learning - I think because of the current attention mechanism that is used by that model (MQA) it is not possible to support prompt learning methods as it involves concatenating the past key values in a specific way that is currently not supported in PEFT. Not sure if the fix should go in PEFT or in transformers though, but for now I would advocate to skip it unless there are a lot of request for that model + prompt learning

**Comment by BenjaminBossan** [comment]
> the current attention mechanism that is used by that model (MQA) it is not possible to support prompt learning methods

I see, thanks for explaining. This also means that I can re-use the `skip_non_pt_mqa` function instead of implementing a new one, I made that change.

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -75,6 +75,22 @@
 }
 
 
+def _get_batch_size(input_ids: Optional[torch.Tensor], inputs_embeds: Optional[torch.Tensor]) -> int:
```
Comment (pacman100) [inline src/peft/peft_model.py:78]: can we move it to utils?

**Review by pacman100 — COMMENTED** [review]
Thank you @BenjaminBossan for fixing the issues, LGTM 🚀. Left a comment.

**Review by pacman100 — APPROVED** [review]
Thank you @BenjaminBossan! 🚀

**Pull request merged**